pkg/plugins/noderesourcetopology/filter.go

```diff
@@ -0,0 +1,128 @@
+package noderesourcetopology
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+
+	topologyv1alpha1 "github.com/gocrane/api/topology/v1alpha1"
+
+	"github.com/gocrane/crane-scheduler/pkg/utils"
+)
+
+const (
+	ErrReasonNUMAResourceNotEnough = "node(s) had insufficient resource of NUMA node"
+	ErrReasonFailedToGetNRT        = "node(s) failed to get NRT"
+)
+
+// PreFilter invoked at the prefilter extension point.
+func (tm *TopologyMatch) PreFilter(
+	ctx context.Context,
+	state *framework.CycleState,
+	pod *corev1.Pod,
+) (*framework.PreFilterResult, *framework.Status) {
+	var indices []int
+	if tm.topologyAwareResources.Has(string(corev1.ResourceCPU)) {
+		indices = GetPodTargetContainerIndices(pod)
+	}
+	resources := computeContainerSpecifiedResourceRequest(pod, indices, tm.topologyAwareResources)
+	state.Write(stateKey, &stateData{
+		aware:                   IsPodAwareOfTopology(pod.Annotations),
+		targetContainerIndices:  indices,
+		targetContainerResource: resources,
+		podTopologyByNode:       make(map[string]*nodeWrapper),
+	})
+	return nil, nil
+}
+
+// PreFilterExtensions returns prefilter extensions, pod add and remove.
+func (tm *TopologyMatch) PreFilterExtensions() framework.PreFilterExtensions {
+	return nil
+}
+
+// Filter will check if there exists numa node which has sufficient resource for the pod.
+func (tm *TopologyMatch) Filter(
+	ctx context.Context,
+	state *framework.CycleState,
+	pod *corev1.Pod,
+	nodeInfo *framework.NodeInfo,
+) *framework.Status {
+	s, err := getStateData(state)
+	if err != nil {
+		return framework.AsStatus(err)
+	}
+
+	if nodeInfo.Node() == nil {
+		return framework.NewStatus(framework.Error, fmt.Sprintf("node(s) not found"))
+	}
+
+	if utils.IsDaemonsetPod(pod) || len(s.targetContainerIndices) == 0 {
+		return nil
+	}
+
+	nrt, err := tm.lister.Get(nodeInfo.Node().Name)
+	if err != nil {
+		return framework.NewStatus(framework.Unschedulable, ErrReasonFailedToGetNRT)
+	}
+	// let kubelet handle cpuset
+	if nrt.CraneManagerPolicy.CPUManagerPolicy != topologyv1alpha1.CPUManagerPolicyStatic {
```

Comment:
The `CPUManagerPolicy` specifies **crane** cpu manager. So only kubelet cpumanager policy is not static can we try to carry on topology scheduling, otherwise kubelet and crane-agent will update CPUSet at same time which may cause a conflict.